File: library/src/main/java/com/github/ohohcakester/algorithms/astar/AStarOctileHeuristic.java
package com.github.ohohcakester.algorithms.astar;

import com.github.ohohcakester.grid.GridGraph;

public class AStarOctileHeuristic extends AStarStaticMemory {
	public AStarOctileHeuristic(GridGraph graph, int sx, int sy, int ex, int ey) {
		super(graph, sx, sy, ex, ey);
	}

	public static AStarOctileHeuristic postSmooth(GridGraph graph, int sx, int sy, int ex, int ey) {
		AStarOctileHeuristic algo = new AStarOctileHeuristic(graph, sx, sy, ex, ey);
		algo.setPostSmoothingOn(true);
		return algo;
	}

	protected float heuristic(int x, int y) {
		return getGraph().octileDistance(x, y, getEx(), getEy());
	}
}

File: baselibrary/src/main/kotlin/com/github/ohohcakester/algorithms/astarstatic/AStarOctileHeuristic.kt
package com.github.ohohcakester.algorithms.astarstatic

import com.github.ohohcakester.grid.GridGraph

class AStarOctileHeuristic(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int) : AStarStaticMemory(graph, sx, sy, ex, ey) {
	companion object {
		fun postSmooth(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int): AStarOctileHeuristic {
			val algo = AStarOctileHeuristic(graph, sx, sy, ex, ey)
			algo.postSmoothingOn = true
			algo.repeatedPostSmooth = false
			return algo
		}

		fun repeatedPostSmooth(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int): AStarOctileHeuristic {
			val algo = AStarOctileHeuristic(graph, sx, sy, ex, ey)
			algo.postSmoothingOn = true
			algo.repeatedPostSmooth = true
			return algo
		}
	}

	override fun heuristic(x: Int, y: Int) = graph.octileDistance(x, y, ex, ey)
}
